### PR TITLE
Development Tooling

### LINKED ISSUE #3: Use perl instead of sed
**Summary:** 

Use `perl` instead of `sed`

**Why do we need this?**

- Predictable behavior across platforms (i.e. `-i` flag)
- More powerful regex
- `perl` is installed by default on Mac and Unix-like operating systems. (https://www.perl.org/get.html)

**What is already there? What do you see now?**

`sed` invocations in scripts and `.make`

**What is missing? What do you want to see?**

`sed` invocations replaced by `perl`

**How do you propose to implement this?**

Replace the `sed` invocations by `perl`

----

Original issue: https://github.com/TheThingsIndustries/lorawan-stack/issues/224 by @rvolosatovs

### LINKED ISSUE #11: Use Mage instead of Make and clean development tooling
**Summary:** 

Our development, build and CI tooling is getting a bit out of hand. Makefiles are complicated and annoying, so we decided to switch to [Mage](https://magefile.org/) which allows us to write our development, build and CI tooling in Go.

**What is already there? What do you see now?**

We have simply written too many Make rules that aren't used, or give unexpected results. The dependency graph has more resemblance to spaghetti than to a directed acyclic graph. Variables are overwritten in multiple places in different files.

**What is missing? What do you want to see?**

I'd like to see a clearer separation in build targets, so that we can split our CI jobs in properly scoped containers:

- Go quality: Testing, Linting, etc.
- JS quality: Testing, Linting, etc.
- Build: Full build with clear dependency flow (first JS, then Go).
- Checking the repo: Regenerate protos, messages, etc. and check for diffs.

The last job ("Checking") is added so that the other jobs can run in parallel. Even though Go tests use `.pb.go` files, we don't need to regenerate them, because we can assume that any changes would already be built and committed to the repo. The "Checking" job will make sure that this is indeed the case before PRs are merged. 

Obviously, the Go tests should not invoke `yarn` or `webpack`, because our Go tests can run just fine without the webui hashes. 

The other way around could be a bit harder, because Mage requires Go. Therefore we may have to find a nice solution for running our JS quality builds in a container without Go.

**How do you propose to implement this?**

For a smooth transition we should probably start migrating Make rules to Mage one by one. The old Make rules can just call Mage until everyone is used to Mage.

I think it's important to consider the differences between development and CI. In development environments, both Go and NodeJS are installed, but on CI this may not be the case, as Go and JS jobs could run in different containers (with different images). The `mage` binary could be built separately of the actual jobs (see below), but it should still avoid calling programs that are specific to one environment (generating protos, calling `node_modules/.bin/json`, etc.). The simplest way to avoid these kind of problems is to avoid dependencies (so Go testing should not depend on the `version/ttn.go` being generated by a task that also calls `node_modules/.bin/json`). We can add more higher-level tasks for development that combine multiple lower-level tasks.

Running JS quality jobs in a node container could be a challenge. Perhaps a multi-stage build with Github Actions instead of Travis could help here. The first action could generate the `.mage/run` binary, which is then used by the subsequent js quality action.

----

Original issue: https://github.com/TheThingsIndustries/lorawan-stack/issues/712 by @htdvisser

### LINKED ISSUE #34: Consider moving to babel 7
**Summary:** 
Babel 7 has been released and there introducing quite many changes that can affect us as well.

* `stage` presets are deprecated. we include `babel-preset-stage-0`.
*  scoped packages, `babel-*` -> `@babel/*`

At the moment the only package requiring the newest babel version is `@storybook/react`, but I can image that later there will be more. This issue is rather low priority and we can using babel 6 for quite a while.

For reference, [the announcement post](https://babeljs.io/blog/2018/08/27/7.0.0).
Especially the major breaking changes, to quote:
> - Drop support for un-maintained Node versions: 0.10, 0.12, 4, 5 (details)
> - Move us to the @babel namespace by switching to using "scoped" packages (details). This helps differentiate official packages, so babel-core becomes @babel/core (and no squatting)
> - Remove (and stop publishing) any yearly presets (preset-es2015, etc) (details). @babel/preset-env replaces the need for these as it includes all yearly additions as well as the ability to target a specific set of browsers
> - Also drop the "Stage" presets (@babel/preset-stage-0, etc) in favor of opting into individual proposals. Similarly remove proposals from @babel/polyfill by default (details). Please consider reading the whole post on this for more explanation.
> - Some packages have renames: any TC39 proposal plugin will now be -proposal instead of -transform (details). So @babel/plugin-transform-class-properties becomes @babel/plugin-proposal-class-properties.
> - Introduce a peerDependency on @babel/core for certain user-facing packages (e.g. babel-loader, @babel/cli, etc) (details)

----

Original issue: https://github.com/TheThingsIndustries/lorawan-stack/issues/1186 by @bafonins

### LINKED ISSUE #66: Move babel configuration to .babelrc
**Summary:** 
We have two babel configurations in `.babelrc` and `package.json` and should get rid of one of them

**What do you see now?**
in `package.json`
```
  "babel": {
    "presets": [
      "env",
      "stage-0",
      "react"
    ],
    "plugins": [
      "transform-decorators-legacy",
      "transform-class-properties",
      "transform-runtime",
      [
        "react-intl-auto",
        {
          "extractComments": true,
          "filebase": true,
          "removePrefix": "pkg.webui"
        }
      ],
      [
        "react-intl",
        {
          "messagesDir": "./.cache/messages",
          "extractSourceLocation": true
        }
      ]
    ]
  },
```

in `.babelrc`
```
{
  "presets": [
    "env",
    "stage-0",
    "react"
  ],
  "plugins": [
    "transform-decorators-legacy",
    "transform-runtime"
  ]
}
```

**What do you want to see instead?**

A single configuration in `.babelrc`

**How do you propose to implement this?**
Move the whole `babel` section from `package.json` to `.babelrc`

**What can you do yourself and what do you need help with?**
Do this

cc @kschiffer 

----

Original issue: https://github.com/TheThingsIndustries/lorawan-stack/issues/1601 by @bafonins

### PR DESCRIPTION
**Pull request by htdvisser**
**Summary:**

This PR migrates our development tooling to Mage.

Closes #3 
Closes #11
Closes #66
Closes #34 

**Changes:**

- Rewrote Git hooks to Mage
- Rewrote headers check to Mage (and removed problematic fix/delete)
- Rewrote Go tooling to Mage
- Rewrote JS tooling to Mage
- Rewrote most of the proto tooling to Mage
- Updated Babel and some related fixes

**Notes for Reviewers:**

There are a lot of fixup commits in here (although they aren't indicated with `!fixup`), but unfortunately rebasing all of this isn't feasible.

The `fix-grpc-gateway-names.sh` script is still TODO. @rvolosatovs promised to solve that in a future PR

### CODE REVIEW

**Comment by kschiffer** [comment]
This now also closes #34.

File: .mage/js.go [hunk 1]
```diff
@@ -60,17 +71,246 @@ func (js Js) yarn() (func(args ...string) error, error) {
...
+		fmt.Println("Running eslint on .snap files")
+	}
+	eslint, err := js.eslint()
+	if err != nil {
+		return err
+	}
+	return eslint("./pkg/webui/**/*.snap", "--no-ignore", "--color")
+}
+
+func (js Js) LintAll() {
```
Comment (bafonins) [inline .mage/js.go]: missing comment

File: pkg/webui/console/api/stubs/fake-data.js [hunk 2]
```diff
@@ -76,13 +76,12 @@ const applicationsApiKeys = applications.reduce(function (acc, curr, idx, apps)
...
 const devices = [ ...new Array(DEVICES_COUNT).keys() ]
   .map(function () {
-    const app = applications[Math.floor(Math.random() * APPLICATIONS_COUNT)]
 
     return {
       ids: {
         device_id: faker.lorem.slug(),
         application_ids: {
-          application_id: app.ids.application_id,
+          application_id: '',
```
[the remark this answers is not in the document]
Comment (kschiffer) [reply pkg/webui/console/api/stubs/fake-data.js]: After the babel 7 upgrade, this line resulted in an error. Didn't want to bother a lot about it, as it is dead code anyway.

File: package.json [hunk 3]
```diff
@@ -90,7 +121,26 @@
...
+      "@babel/plugin-proposal-optional-chaining",
+      [
+        "@babel/plugin-proposal-pipeline-operator",
+        {
+          "proposal": "minimal"
+        }
+      ],
+      "@babel/plugin-proposal-nullish-coalescing-operator",
+      "@babel/plugin-proposal-do-expressions",
+      "@babel/plugin-proposal-function-bind"
```
[the remark this answers is not in the document]
Comment (kschiffer) [reply package.json]: We had them enabled before, that's why they're added now individually. I can also remove some, as it's true that we actually don't use most of them.

**Comment by kschiffer** [comment]
This now also closes #66.

**Comment by kschiffer** [comment]
What's the status here?

**Comment by htdvisser** [comment]
Nothing changed. Please continue working on it, @kschiffer.

Also, @rvolosatovs needs to start looking into the proto stuff

**Comment by kschiffer** [comment]
I have added the missing targets for the JS SDK proto and api definition generation and replaced the respective make targets with it. I've also updated target dependencies, so that `mage js:build` will run anything needed to build the frontend from scratch, including SDK build processes.

@bafonins can you double check if this works for you?

**Comment by htdvisser** [comment]
@TheThingsNetwork/stack-developers I think we're pretty much done here. I just did some final cleanup and ran through the DEVELOPMENT.md file to make sure we didn't break too much. 

I'm sure we'll find some more mistakes after this is merged, and will make more changes later, but let's first get this thing in.

File: pkg/webui/components/navigation/side/item/__snapshots__/index_test.js.snap [hunk 4]
```diff
@@ -23,7 +23,7 @@ exports[`SideNavigationItem is collapsable should match snapshot 1`] = `
       ]
     }
     onClick={[Function]}
-    onExpand={[Function]}
+    onExpand={[MockFunction]}
```
Comment (bafonins) [inline pkg/webui/components/navigation/side/item/__snapshots__/index_test.js.snap]: what happened here?

File: package.json [hunk 5]
```diff
@@ -62,37 +64,10 @@
     "webpack-cli": "^3.1.2",
     "webpack-dev-server": "^3.1.10",
     "webpack-plugin-hash-output": "^3.1.0",
-    "webpack-shell-plugin": "0.4.5",
+    "webpack-shell-plugin": "https://github.com/cdeutsch/webpack-shell-plugin.git#bee537d",
```
Comment (bafonins) [inline package.json]: doesnt look right

File: config/eslintrc.yaml [hunk 6]
```diff
@@ -17,6 +17,11 @@ globals:
   mount: true
   snapshotDiff: true
 
+parserOptions:
+  ecmaFeatures:
+    # Remove after migration
+    legacyDecorators: true
```
Comment (bafonins) [inline config/eslintrc.yaml]: can we remove this?

File: .mage/js.go [hunk 7]
```diff
@@ -60,17 +71,265 @@ func (js Js) yarn() (func(args ...string) error, error) {
...
+// Serve builds necessary bundles and serves the console for development.
+func (js Js) Serve() {
+	mg.Deps(js.ServeMain)
+}
+
+// ServeMain runs webpack-dev-server
+func (js Js) ServeMain() error {
+	mg.Deps(js.Translations, js.BackendTranslations, js.BuildDll)
+	if mg.Verbose() {
+		fmt.Println("Running Webpack for Main Bundle in watch mode…")
```
Comment (bafonins) [inline .mage/js.go]: I think we should always display this message. At the moment, after running `mage js:serve` there is no output at all while the terminal is blocked by the `webpack-dev-server` process. the only reason to use this is for development purposes and should not bother developers who just want to deploy the stack and keep output clean.

File: .mage/js.go [hunk 8]
```diff
@@ -60,17 +71,265 @@ func (js Js) yarn() (func(args ...string) error, error) {
...
+
+// LintAll runs linters over js and snap files.
+func (js Js) LintAll() {
+	mg.Deps(js.Lint, js.LintSnap)
+}
+
+// Storybook runs a local server with storybook.
+func (js Js) Storybook() error {
+	if mg.Verbose() {
+		fmt.Println("Serving storybook…")
```
Comment (bafonins) [inline .mage/js.go]: same as https://github.com/TheThingsNetwork/lorawan-stack/pull/377/files#r278934042

File: pkg/webui/components/navigation/side/item/__snapshots__/index_test.js.snap (hunk 4 above)
Comment (kschiffer) [reply pkg/webui/components/navigation/side/item/__snapshots__/index_test.js.snap]: I needed to update jest to make it work with babel 7, which seemed to have changed some stuff in snapshot creation. I updated the snapshots to make the tests pass again.

File: package.json (hunk 5 above)
Comment (kschiffer) [reply package.json]: [See this](https://github.com/1337programming/webpack-shell-plugin/issues/61#issuecomment-384090619). It's the only way to get rid of the 
```
(node:83992) DeprecationWarning: Tapable.plugin is deprecated. Use new API on `.hooks` instead
    at WebpackShellPlugin.apply (/projectDir/node_modules/webpack-shell-plugin/lib/index.js:236:16
```
deprecation warning. The original `webpack-shell-plugin` appears discontinued. I don't know if we should rather live with the deprecation warning.

File: config/eslintrc.yaml (hunk 6 above)
Comment (kschiffer) [reply config/eslintrc.yaml]: If we do, then our linters would throw warnings for `@decorators` in front of `export` statements (which is deprecated since babel 7). My suggestion would be to add this rule to allow the legacy syntax now and migrate the codebase later when we have time.

File: .mage/js.go (hunk 7 above)
Comment (kschiffer) [reply .mage/js.go]: I agree. Had a discussion about this as well with @htdvisser, who said (correct me if I'm wrong) that we should for now rather follow the convention and show target output only when mage's verbose mode `-v` is enabled. Overall the output of targets is not really consistent yet. We should look into this once more.

**Comment by kschiffer** [comment]
> However, can we avoid multiple calls to `make js.dev-deps`?

I've added a couple of more src/dist checks to reduce unnecessary reruns. Also for `jsSdk:link`.

File: .mage/js.go (hunk 7 above)
Comment (kschiffer) [reply .mage/js.go]: I have piped the output for these serve targets back to `os.Stdout` by default. @htdvisser feel free to drop `51dc0bf` if you disagree.

File: .mage/proto.go [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+// JsSDKClean removes generated javascript SDK protos.
+func (p Proto) JsSDKClean(context.Context) error {
+	return sh.Rm(filepath.Join("sdk", "js", "generated", "api.json"))
+}
+
+// All generates protos.
+func (p Proto) All(ctx context.Context) {
+	mg.CtxDeps(ctx, Proto.Image)
```
Comment (rvolosatovs) [inline .mage/proto.go]: Remove this.
1. It's redundant, since `docker run` will automatically pull the image if not available.
2. It breaks `proto:all` in offline scenarios in cases where the image is already available and in cases where image is available locally, but not in the repository.
3. This target generates protos. Updating the image is an unexpected side-effect.

Comment (htdvisser) [reply .mage/proto.go]: When I ran this `proto:all` without pulling the image first, it started pulling the image concurrently for all the jobs below. We don't want that. Feel free to suggest an alternative.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -245,7 +245,7 @@ In the API, the enum descriptions, error messages and event descriptions availab
 These messages are then collected in the `config/messages.json` file, which will be processed in the frontend build process, but may also be used by other (native) user interfaces. When you define new enums, errors or events or when you change them, the messages need to be updated into the `config/messages.json` file.
 
 ```sh
-make messages
+make go.messages
```
Comment (rvolosatovs) [inline CONTRIBUTING.md]: Why not `go:messages`?

File: DEVELOPMENT.md [hunk 11]
```diff
@@ -142,7 +143,7 @@ From the `.proto` files, we generate code using the `protoc` compiler. As we pla
 The actual commands for compilation are handled by our Makefile, so the only thing you have to execute, is:
 
 ```sh
-make protos.clean protos
+make proto.clean proto.all
```
Comment (rvolosatovs) [inline DEVELOPMENT.md]: Why not `proto:clean` and `proto:all` ?

File: Makefile [hunk 12]
```diff
@@ -45,18 +64,10 @@ docs:
...
-	rm -rf dist
-
-translations: messages js.translations
+.PHONY: git.diff
+git.diff:
+	@if [[ ! -z "`git diff`" ]]; then \
+		$(err) "Previous operations have created changes that were not recorded in the repository. Please make those changes on your local machine before pushing them to the repository:"; \
+		git diff; \
+		exit 1; \
+	fi
```
Comment (rvolosatovs) [inline Makefile]: Why did you remove the passthrough target to `mage`?

File: .mage/rules.make [hunk 13]
```diff
@@ -0,0 +1,257 @@
+# Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
```
Comment (rvolosatovs) [inline .mage/rules.make]: I thought we're migrating to Mage in this PR? Why do we need these?

File: Makefile [hunk 14]
```diff
@@ -12,30 +12,49 @@
...
+.PHONY: build
+build:
+	@echo "For development:"
+	@echo "  run \"make $(BUILD_RULES)\""
+	@echo "  and then \"go run ./cmd/ttn-lw-stack\" or \"go run ./cmd/ttn-lw-cli\"."
+	@echo "For production:"
+	@echo "  run \"make build-all\""
+
+.PHONY: build-all
+build-all:
```
Comment (rvolosatovs) [inline Makefile]: Why is this still here?

File: Makefile [hunk 15]
```diff
@@ -12,30 +12,49 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
```
Comment (rvolosatovs) [inline Makefile]: Why do we still have targets here? Aren't we migrating to `mage`?

File: .mage/proto.go (hunk 9 above)
Comment (johanstokking) [reply .mage/proto.go]: I favor offline support over concurrent pulling images. If we can have (explicit) offline support, I'm fine with that too, but we need offline support.

Comment (rvolosatovs) [reply .mage/proto.go]: Fixed that in https://github.com/TheThingsNetwork/lorawan-stack/pull/377/commits/82c670f81398aaea1d8b7b740543353ab0c7922b

**Comment by bafonins** [comment]
```
export NODE_ENV=production
make init deps
mage js:build
```
gives:
```
Error: Couldn't find preset "ttn" relative to directory "$GOPATH/src/go.thethings.network/lorawan-stack/sdk/js"
    at /usr/local/lib/node_modules/babel-cli/node_modules/babel-core/lib/transformation/file/options/option-manager.js:293:19
    at Array.map (<anonymous>)
    at OptionManager.resolvePresets (/usr/local/lib/node_modules/babel-cli/node_modules/babel-core/lib/transformation/file/options/option-manager.js:275:20)
    at OptionManager.mergePresets (/usr/local/lib/node_modules/babel-cli/node_modules/babel-core/lib/transformation/file/options/option-manager.js:264:10)
    at OptionManager.mergeOptions (/usr/local/lib/node_modules/babel-cli/node_modules/babel-core/lib/transformation/file/options/option-manager.js:249:14)
    at OptionManager.init (/usr/local/lib/node_modules/babel-cli/node_modules/babel-core/lib/transformation/file/options/option-manager.js:368:12)
    at File.initOptions (/usr/local/lib/node_modules/babel-cli/node_modules/babel-core/lib/transformation/file/index.js:212:65)
    at new File (/usr/local/lib/node_modules/babel-cli/node_modules/babel-core/lib/transformation/file/index.js:135:24)
    at Pipeline.transform (/usr/local/lib/node_modules/babel-cli/node_modules/babel-core/lib/transformation/pipeline.js:46:16)
    at transform (/usr/local/lib/node_modules/babel-cli/lib/babel/util.js:50:22)
error Command failed with exit code 1.                                                           
Error: running "node_modules/.bin/yarn --cwd=sdk/js run build" failed with exit code 1
```

File: .mage/translations.js [hunk 16]
```diff
@@ -202,7 +202,7 @@ const readMessages = async function () {
 }
 
 /**
- * Read and parse (and marshal) the backend messages, coming from `make go.messages`
+ * Read and parse (and marshal) the backend messages, coming from `./mage go.messages`
```
Comment (johanstokking) [inline .mage/translations.js]: colon

Comment (htdvisser) [reply .mage/translations.js]: :sob:

**Comment by htdvisser** [comment]
@rvolosatovs we got rid of the proxy, now we run `make init` to set up Mage, after which we run `./mage <rules...>`.

**Comment by htdvisser** [comment]
Notes to all: 

- `.make/dev.make` will be rewritten to Mage in a later PR
- `.mage/scripts/fix-grpc-gateway-names.sh` will be rewritten to Mage in a later PR
